providers/freebsd/memory_freebsd_cgo.go

```diff
@@ -0,0 +1,114 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build freebsd && cgo
+
+package freebsd
+
+// #cgo LDFLAGS: -lkvm
+//#include <sys/cdefs.h>
+//#include <sys/types.h>
+//#include <sys/sysctl.h>
+
+//#include <paths.h>
+//#include <kvm.h>
+//#include <stdlib.h>
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	hwPhysmemMIB         = "hw.physmem"
+	hwPagesizeMIB        = "hw.pagesize"
+	vmVmtotalMIB         = "vm.vmtotal"
+	vmSwapmaxpagesMIB    = "vm.swap_maxpages"
+	vfsNumfreebuffersMIB = "vfs.numfreebuffers"
+	devNull              = "/dev/null"
+	kvmOpen              = "kvm_open"
+)
+
+func PageSize() (uint32, error) {
+	var pageSize uint32
+	if err := sysctlByName(hwPagesizeMIB, &pageSize); err != nil {
+		return 0, fmt.Errorf("failed to get hw.pagesize: %w", err)
+	}
+
+	return pageSize, nil
+}
+
+func SwapMaxPages() (uint32, error) {
+	var maxPages uint32
+	if err := sysctlByName(hwPhysmemMIB, &maxPages); err != nil {
+		return 0, fmt.Errorf("failed to get vm.swap_maxpages: %w", err)
+	}
+
+	return maxPages, nil
+}
+
+func TotalMemory() (uint64, error) {
+	var size uint64
+	if err := sysctlByName(hwPhysmemMIB, &size); err != nil {
+		return 0, fmt.Errorf("failed to get hw.physmem: %w", err)
+	}
+
+	return size, nil
+}
+
+func VmTotal() (vmTotal, error) {
+	var vm vmTotal
+	if err := sysctlByName(vmVmtotalMIB, &vm); err != nil {
+		return vmTotal{}, fmt.Errorf("failed to get vm.vmtotal: %w", err)
+	}
+
+	return vm, nil
+}
+
+func NumFreeBuffers() (uint32, error) {
+	var numfreebuffers uint32
+	if err := sysctlByName(vfsNumfreebuffersMIB, &numfreebuffers); err != nil {
+		return 0, fmt.Errorf("failed to get vfs.numfreebuffers: %w", err)
+	}
+
+	return numfreebuffers, nil
+}
+
+func KvmGetSwapInfo() (kvmSwap, error) {
+	var kdC *C.struct_kvm_t
+
+	devNullC := C.CString(devNull)
+	defer C.free(unsafe.Pointer(devNullC))
+	kvmOpenC := C.CString(kvmOpen)
+	defer C.free(unsafe.Pointer(kvmOpenC))
+
+	if kdC, err := C.kvm_open(nil, devNullC, nil, unix.O_RDONLY, kvmOpenC); kdC == nil {
```

Comment:
This uses `kdC, err := ` so it is shadowing the definition on L95. This means that the `kvm_close` on L106 is not operating on the value from `kvm_open`.

Reply:
3952336